src/cargo/util/rustc.rs

```diff
@@ -62,4 +75,157 @@ impl Rustc {
+                        }
+                    }
+                    Err(e) => {
+                        info!("failed to read rustc info cache: {}", e);
+                        empty
+                    }
+                };
+                return Cache {
+                    cache_location: Some(cache_location),
+                    dirty,
+                    data,
+                };
+
+                fn read(path: &Path) -> CargoResult<CacheData> {
+                    let json = paths::read(path)?;
+                    Ok(serde_json::from_str(&json)?)
+                }
+            }
+            (_, fingerprint) => {
+                if let Err(e) = fingerprint {
+                    warn!("failed to calculate rustc fingerprint: {}", e);
+                }
+                info!("rustc info cache disabled");
+                Cache {
+                    cache_location: None,
+                    dirty: false,
+                    data: CacheData::default(),
+                }
+            }
+        }
+    }
+
+    fn cached_output(&mut self, cmd: &ProcessBuilder) -> CargoResult<(String, String)> {
+        let calculate = || {
+            let output = cmd.exec_with_output()?;
+            let stdout = String::from_utf8(output.stdout)
+                .map_err(|_| internal("rustc didn't return utf8 output"))?;
+            let stderr = String::from_utf8(output.stderr)
+                .map_err(|_| internal("rustc didn't return utf8 output"))?;
+            Ok((stdout, stderr))
+        };
+        if self.cache_location.is_none() {
+            info!("rustc info uncached");
+            return calculate();
+        }
+
+        let key = process_fingerprint(cmd);
+        match self.data.outputs.entry(key) {
+            Entry::Occupied(entry) => {
+                info!("rustc info cache hit");
+                Ok(entry.get().clone())
+            }
+            Entry::Vacant(entry) => {
+                info!("rustc info cache miss");
+                let output = calculate()?;
+                entry.insert(output.clone());
+                self.dirty = true;
+                Ok(output)
+            }
+        }
+    }
+}
+
+impl Drop for Cache {
+    fn drop(&mut self) {
+        match (&self.cache_location, self.dirty) {
+            (&Some(ref path), true) => {
+                let json = serde_json::to_string(&self.data).unwrap();
+                match paths::write(path, json.as_bytes()) {
+                    Ok(()) => info!("updated rustc info cache"),
+                    Err(e) => warn!("failed to update rustc info cache: {}", e),
+                }
+            }
+            _ => (),
+        }
+    }
+}
+
+fn rustc_fingerprint(path: &Path) -> CargoResult<u64> {
+    let mut hasher = SipHasher::new_with_keys(0, 0);
+
+    let path = paths::resolve_executable(path)?;
+    path.hash(&mut hasher);
+
+    paths::mtime(&path)?.hash(&mut hasher);
+
+    match (env::var("RUSTUP_HOME"), env::var("RUSTUP_TOOLCHAIN")) {
```

Comment:
Could you insert a comment here as to what's going on?

Reply:
Done!